### PR TITLE
Action Controller guide edits for grammar and clarity

### PR DESCRIPTION
**Pull request by cantino**
This PR updates the Action Controller guide with some minor grammar and clarity improvements.

### CODE REVIEW

File: guides/source/action_controller_overview.md [hunk 1]
```diff
@@ -143,15 +141,15 @@ So for example, if you are sending this JSON content:
 { "company": { "name": "acme", "address": "123 Carrot Street" } }
 ```
 
-You'll get `params[:company]` as `{ "name" => "acme", "address" => "123 Carrot Street" }`.
+Your controller will receive `params[:company]` as `{ "name" => "acme", "address" => "123 Carrot Street" }`.
 
-Also, if you've turned on `config.wrap_parameters` in your initializer or calling `wrap_parameters` in your controller, you can safely omit the root element in the JSON parameter. The parameters will be cloned and wrapped in the key according to your controller's name by default. So the above parameter can be written as:
+Also, if you've turned on `config.wrap_parameters` in your initializer or called `wrap_parameters` in your controller, you can safely omit the root element in the JSON parameter. In this case, by default, the parameters will be cloned and wrapped with a key chosen based on your controller's name. So the above JSON POST can be written as:
```
Comment (georgeclaghorn) [inline guides/source/action_controller_overview.md]: Wouldn't say "by default," since you're explicitly overriding the default behavior.

File: guides/source/action_controller_overview.md [hunk 2]
```diff
@@ -63,7 +63,7 @@ The [Layouts & Rendering Guide](layouts_and_rendering.html) explains this in mor
 
 `ApplicationController` inherits from `ActionController::Base`, which defines a number of helpful methods. This guide will cover some of these, but if you're curious to see what's in there, you can see all of them in the API documentation or in the source itself.
 
-Only public methods are callable as actions. It is a best practice to lower the visibility of methods which are not intended to be actions, like auxiliary methods or filters.
+Only public methods are callable as actions. It is a best practice to lower the visibility of methods (with `private`) which are not intended to be actions, like auxiliary methods or filters.
```
Comment (georgeclaghorn) [inline guides/source/action_controller_overview.md]: If we're going to be specific, we should also mention that `protected` is an acceptable level of visibility for auxiliary methods.

Comment (cantino) [inline guides/source/action_controller_overview.md]: Okay, I will add that.

File: guides/source/action_controller_overview.md (hunk 1 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: Agreed, will fix shortly.

**Comment by cantino** [comment]
Thanks @georgeclaghorn, I agree with both of those comments and have updated the PR.

**Comment by georgeclaghorn** [comment]
:+1: You'll also want to squash your commits.

**Comment by cantino** [comment]
:heavy_check_mark: Done!

File: guides/source/action_controller_overview.md [hunk 3]
```diff
@@ -104,21 +104,19 @@ end
...
+NOTE: The actual URL in this example will be encoded as "/clients?ids%5b%5d=1&ids%5b%5d=2&ids%5b%5d=3" as the "[" and "]" characters are not allowed in URLs. Most of the time you don't have to worry about this because the browser will encode it for you, and Rails will decode it automatically, but if you ever find yourself having to send those requests to the server manually you should keep this in mind.
 
 The value of `params[:ids]` will now be `["1", "2", "3"]`. Note that parameter values are always strings; Rails makes no attempt to guess or cast the type.
 
-NOTE: Values such as `[nil]` or `[nil, nil, ...]` in `params` are replaced
-with `[]` for security reasons by default. See [Security Guide](security.html#unsafe-query-generation)
-for more information.
+NOTE: Values such as `[nil]` or `[nil, nil, ...]` in `params` are replaced with `[]` for security reasons by default. See [Security Guide](security.html#unsafe-query-generation) for more information.
```
Comment (robin850) [inline guides/source/action_controller_overview.md]: Can you revert this change please ? Even though this seems inconsistent throughout the file, we are slowly wrapping new additions around 80 chars.

File: guides/source/action_controller_overview.md [hunk 4]
```diff
@@ -131,11 +129,11 @@ To send a hash you include the key name inside the brackets:
...
 
-Note that the `params` hash is actually an instance of `ActiveSupport::HashWithIndifferentAccess`, which acts like a hash but lets you use symbols and strings interchangeably as keys.
+The `params` hash is actually an instance of `ActiveSupport::HashWithIndifferentAccess`, which acts like a Hash but lets you use symbols and strings interchangeably as keys.
 
 ### JSON parameters
 
-If you're writing a web service application, you might find yourself more comfortable accepting parameters in JSON format. If the "Content-Type" header of your request is set to "application/json", Rails will automatically convert your parameters into the `params` hash, which you can access as you would normally.
+If you're writing a web service application, you might find yourself more comfortable accepting parameters in JSON format. If the "Content-Type" header of your request is set to "application/json", Rails will automatically load your parameters into the `params` hash, which you can access as you would normally.
```
Comment (robin850) [inline guides/source/action_controller_overview.md]: I guess `convert` was right here ; the process described here concerns the translation/conversion of JSON inside the Ruby land. What do you think ?

File: guides/source/action_controller_overview.md [hunk 5]
```diff
@@ -159,17 +157,17 @@ And assume that you're sending the data to `CompaniesController`, it would then
...
+The `params` hash will always contain the `:controller` and `:action` keys, but you should use the methods `controller_name` and `action_name` instead to access these values. Any other parameters defined by the routing, such as `:id`, will also be available. As an example, consider a listing of clients where the list can show either active or inactive clients. We can add a route which captures the `:status` parameter in a "pretty" URL:
 
 ```ruby
 get '/clients/:status' => 'clients#index', foo: 'bar'
 ```
 
-In this case, when a user opens the URL `/clients/active`, `params[:status]` will be set to "active". When this route is used, `params[:foo]` will also be set to "bar" just like it was passed in the query string. In the same way `params[:action]` will contain "index".
+In this case, when a user opens the URL `/clients/active`, `params[:status]` will be set to "active". When this route is used, `params[:foo]` will also be set to "bar", as if it were passed in via the query string. You will also receive `params[:action]` as "index" and `params[:controller]` as "clients".
```
Comment (robin850) [inline guides/source/action_controller_overview.md]: What do you think about writing "as if it were passed to the query string" ? The combination of "in" with "via" doesn't read that well even though this seems grammatically correct.

File: guides/source/action_controller_overview.md [hunk 6]
```diff
@@ -183,21 +181,21 @@ class ApplicationController < ActionController::Base
 end
 ```
 
-These options will be used as a starting point when generating URLs, so it's possible they'll be overridden by the options passed in `url_for` calls.
+These options will be used as a starting point when generating URLs, so it's possible they'll be overridden by the options passed in to `url_for` calls.
```
Comment (robin850) [inline guides/source/action_controller_overview.md]: Ditto - "passed to `url_for` calls" ?

File: guides/source/action_controller_overview.md (hunk 3 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: Yup, I'll revert this.

File: guides/source/action_controller_overview.md (hunk 4 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: I feel like the action really is taking the JSON data and `loading` it into the params hash.  However, I don't feel strongly and can revert if you prefer.

File: guides/source/action_controller_overview.md (hunk 5 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: Changing it to "as if it were passed in the query string".  A compromise!

File: guides/source/action_controller_overview.md (hunk 6 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: Will change.

**Comment by cantino** [comment]
Thanks @robin850, I've incorporated your suggestions and squashed.

File: guides/source/action_controller_overview.md (hunk 4 above)
Comment (robin850) [inline guides/source/action_controller_overview.md]: Yep, definitely actually since the data is stored in the hash, sorry! :+1: Moreover, there is a `JSON.load` method.

File: guides/source/action_controller_overview.md [hunk 7]
```diff
@@ -143,15 +143,15 @@ So for example, if you are sending this JSON content:
...
 
-Also, if you've turned on `config.wrap_parameters` in your initializer or calling `wrap_parameters` in your controller, you can safely omit the root element in the JSON parameter. The parameters will be cloned and wrapped in the key according to your controller's name by default. So the above parameter can be written as:
+Also, if you've turned on `config.wrap_parameters` in your initializer or called `wrap_parameters` in your controller, you can safely omit the root element in the JSON parameter. In this case, the parameters will be cloned and wrapped with a key chosen based on your controller's name. So the above JSON POST can be written as:
 
 ```json
 { "name": "acme", "address": "123 Carrot Street" }
 ```
 
-And assume that you're sending the data to `CompaniesController`, it would then be wrapped in `:company` key like this:
+And, assuming that you're sending the data to `CompaniesController`, it would then be wrapped with a `:company` key like this:
```
Comment (robin850) [inline guides/source/action_controller_overview.md]: What do you think about "within the `:company` key" ?

File: guides/source/action_controller_overview.md [hunk 8]
```diff
@@ -183,21 +183,21 @@ class ApplicationController < ActionController::Base
...
 
 ### Strong Parameters
 
 With strong parameters, Action Controller parameters are forbidden to
 be used in Active Model mass assignments until they have been
-whitelisted. This means you'll have to make a conscious choice about
-which attributes to allow for mass updating and thus prevent
-accidentally exposing that which shouldn't be exposed.
+whitelisted. This means that you'll have to make a conscious decision about
+which attributes to allow for mass update. This is a best security
```
Comment (robin850) [inline guides/source/action_controller_overview.md]: s/best/better

Comment (cantino) [inline guides/source/action_controller_overview.md]: "best practice" is an industry term.  That said, I'm happy to change it if you prefer.

File: guides/source/action_controller_overview.md (hunk 7 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: So, "it would then be wrapped within the `:company` key"?  Sure, I can do that.

Comment (cantino) [inline guides/source/action_controller_overview.md]: Done.

File: guides/source/action_controller_overview.md (hunk 8 above)
Comment (matthewd) [inline guides/source/action_controller_overview.md]: "best X practice" sounds very odd to me. I think "X best practice" would be the more usual spelling. This whole sentence seems not-so-great at the moment, though: "help prevent accidentally allowing"?

I don't have a concrete improvement to suggest, but I do think there's one to be found.

File: guides/source/action_controller_overview.md [hunk 9]
```diff
@@ -183,21 +183,21 @@ class ApplicationController < ActionController::Base
...
+practice to help prevent accidentally allowing users to update sensitive
+model attributes.
 
-In addition, parameters can be marked as required and flow through a
-predefined raise/rescue flow to end up as a 400 Bad Request with no
-effort.
+In addition, parameters can be easily marked as required and will flow through a
```
Comment (matthewd) [inline guides/source/action_controller_overview.md]: easily :-1:

File: guides/source/action_controller_overview.md (hunk 8 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: Okay, changing it :)

File: guides/source/action_controller_overview.md (hunk 9 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: Okay

File: guides/source/action_controller_overview.md [hunk 10]
```diff
@@ -21,11 +21,11 @@ After reading this guide, you will know:
...
+Action Controller is the C in MVC. After routing has determined which controller to use for a request, the controller is responsible for making sense of the request and producing the appropriate output. Luckily, Action Controller does most of the groundwork for you and uses smart conventions to make this as straightforward as possible.
 
 For most conventional [RESTful](http://en.wikipedia.org/wiki/Representational_state_transfer) applications, the controller will receive the request (this is invisible to you as the developer), fetch or save data from a model and use a view to create HTML output. If your controller needs to do things a little differently, that's not a problem, this is just the most common way for a controller to work.
 
-A controller can thus be thought of as a middle man between models and views. It makes the model data available to the view so it can display that data to the user, and it saves or updates data from the user to the model.
+A controller can thus be thought of as a middle-man between models and views. It makes the model data available to the view so it can display that data to the user, and it saves or updates data from the user to the model.
```
Comment (trosborn) [inline guides/source/action_controller_overview.md]: I have only ever seen middleman written as a single word, never hyphenated. The OED also only spells it as a single word, so I assume this is not just an American English convention. :smile:

Comment (cantino) [inline guides/source/action_controller_overview.md]: Agreed.  Just changed those two.

**Comment by cantino** [comment]
I've rebased and addressed all comments.  Is this ready to merge?

**Comment by trosborn** [comment]
Other than that, every LGTM!

File: guides/source/action_controller_overview.md [hunk 11]
```diff
@@ -131,11 +131,11 @@ To send a hash you include the key name inside the brackets:
 
 When this form is submitted, the value of `params[:client]` will be `{ "name" => "Acme", "phone" => "12345", "address" => { "postcode" => "12345", "city" => "Carrot City" } }`. Note the nested hash in `params[:client][:address]`.
 
-Note that the `params` hash is actually an instance of `ActiveSupport::HashWithIndifferentAccess`, which acts like a hash but lets you use symbols and strings interchangeably as keys.
+The `params` hash is actually an instance of `ActiveSupport::HashWithIndifferentAccess`, which acts like a Hash but lets you use symbols and strings interchangeably as keys.
```
Comment (rafaelfranca) [inline guides/source/action_controller_overview.md]: This is not true true anymore. `params` is an instance of `ActionController::Parameters`. Maybe we should not talk about which class it inherit but only that symbols and strings are interchangeable

File: guides/source/action_controller_overview.md [hunk 12]
```diff
@@ -183,21 +183,21 @@ class ApplicationController < ActionController::Base
...
-accidentally exposing that which shouldn't be exposed.
+whitelisted. This means that you'll have to make a conscious decision about
+which attributes to allow for mass update. This is a better security
+practice to help prevent accidentally allowing users to update sensitive
+model attributes.
 
-In addition, parameters can be marked as required and flow through a
-predefined raise/rescue flow to end up as a 400 Bad Request with no
-effort.
+In addition, parameters can be  marked as required and will flow through a
```
Comment (rafaelfranca) [inline guides/source/action_controller_overview.md]: extra space after the `be`

Comment (cantino) [inline guides/source/action_controller_overview.md]: Fixed

File: guides/source/action_controller_overview.md (hunk 11 above)
Comment (cantino) [inline guides/source/action_controller_overview.md]: Great point.  Changed to

```
The `params` object acts like a Hash, but lets you use symbols and strings interchangeably as keys.
```

**Comment by cantino** [comment]
Additional changes made.  Thanks @trosborn & @rafaelfranca!

**Comment by rafaelfranca** [comment]
Awesome work!

**Comment by cantino** [comment]
Thanks!
